### DISCUSSION TITLE
Retrieve custom variables from ServiceNow Task

### DISCUSSION THREAD

**Discussion by emarzanoBNY** [Snow-Shell/servicenow-powershell#138] [Q&A] [answered]
Hi There, 

I'm using Get-ServiceNowRecord to retrieve all the fields for a task.   This works great for all the default fields, but I don't see any of my custom variables in there.  

Is there anyway to retrieve my custom variables using Get-ServiceNowRecord?  

Thanks for your help!

**Comment by gdbarron** [comment, marked as answer]
Hi @emarzanoBNY.  I've put something together to grab these values.  Do you mind putting it through it's paces?  The branch is https://github.com/Snow-Shell/servicenow-powershell/tree/get-custom-variables and the new property is `-IncludeCustomVariable` on `Get-ServiceNowRecord`.  I've tested on PSv5, but not 7 yet, and working well.

**Reply by emarzanoBNY** [reply]
I will try it out today!

**Reply by emarzanoBNY** [reply]
Get-ServiceNowRecord -Table task -IncludeCustomVariable -Filter @('number', '-eq', 'MYTASKNUMBER') 

executes without error, but contains the same values as  

Get-ServiceNowRecord -Table task -Filter @('number', '-eq', 'MYTASKNUMBER') 

Am I missing something?   How do the custom variables appear in your output?

**Reply by gdbarron** [reply]
Are you sure that task has custom variables?  During my testing, some of the results had custom vars and some didn't so it might have looked like it wasn't working.  I submitted a requested item I knew had values I needed to provide.  RITM000001 doesn't have any custom vars, but the other 2 do.
![image](https://user-images.githubusercontent.com/11862024/121379837-68379700-c912-11eb-98fc-4c1f42cf3e0a.png)

With regards to how they appear, the variables. in the output are the custom ones.  Do you like adding them as typical properties or should we look to have a property called CustomVariable and list them there?

**Reply by emarzanoBNY** [reply]
I believe I do have custom variables in the ticket I am referencing.  I am very new to ServiceNow, so I'm learning as I go.  

Sanitized to share, here is what I see at the bottom of my task in the GUI.  

![image](https://user-images.githubusercontent.com/82045395/121391868-d4b79380-c91c-11eb-8f16-175e62030870.png)

I'm jamming the ticket info into a 

$Ticket = Get-ServiceNowRecord -Table task -IncludeCustomVariable -Filter @('number', '-eq', 'MYTASKNUMBER')

I'm hoping to extract things like the New Hire Name from $ticket.variables.new_hire_name to use in an automatic user provisioning script.  

When I look at the contents of $ticket, I see all the normal fields without any of the variables.

**Reply by gdbarron** [reply]
Can you run the command with `-verbose` and provide me the output, ensuring sensitive data is sanitized?  There are additional api calls being made and I'd like to be sure those look correct because as usual with all developers...it works on my machine 😄

**Reply by emarzanoBNY** [reply]
Sure!  Here's what we're looking at: 

VERBOSE: New-ServiceNowQuery - Advanced
VERBOSE: {
    "Credential":  {
                       "UserName":  "myusername",
                       "Password":  {
                                        "Length":  16
                                    }
                   },
    "Method":  "Get",
    "ContentType":  "application/json",
    "Uri":  "https://mysite.service-now.com/api/now/table/task",
    "Body":  {
                 "sysparm_display_value":  "true",
                 "sysparm_limit":  10,
                 "sysparm_query":  "number=MYTASKNUMBER^ORDERBYDESCopened_at"
             }
}
VERBOSE: GET https://mysite.service-now.com/api/now/table/task?sysparm_display_value=true&sysparm_limit=10&sysparm_query=number%3DMYTASKNUMBER^ORDERBYDESCopened_at with 0-byte payload
VERBOSE: received -1-byte response of content type application/json;charset=UTF-8
VERBOSE: Total number of records for this query: 1
VERBOSE: New-ServiceNowQuery - Advanced
VERBOSE: {
    "Credential":  {
                       "UserName":  "myusername",
                       "Password":  {
                                        "Length":  16
                                    }
                   },
    "Method":  "Get",
    "ContentType":  "application/json",
    "Uri":  "https://mysite.service-now.com/api/now/table/sc_item_option_mtom",
    "Body":  {
                 "sysparm_display_value":  "true",
                 "sysparm_limit":  1000,
                 "sysparm_fields":  "sc_item_option.item_option_new.name,sc_item_option.item_option_new.sys_name,sc_item_option.item_option_new.type",
                 "sysparm_query":  "request_item=68319bf81bacb8d0752843b4bd4bcbf3^sc_item_option.item_option_new.typeIN1,2,3,4,5,6,7,8,9,10,16,18,21,22^ORDERBYDESCopened_at"
             }
}
VERBOSE: GET https://mysite.service-now.com/api/now/table/sc_item_option_mtom?sysparm_display_value=true&sysparm_limit=1000&sysparm_fields=sc_item_option.item_option_new.name%2Csc_item_option.item_option_new.sys_name%2Csc_item_option.item_option_new.type&sysparm_query=request_item%3D68319bf81bacb8d0752843b4bd4bcbf3^sc_item_option.item_option_new.typeIN1%2C2%2C3%2C4%2C5%2C6%2C7%2C8%2C9%2C10%2C16%2C18%2C21%2C22^ORDERBYDESCopened_at with 0-byte payload
VERBOSE: received -1-byte response of content type application/json;charset=UTF-8
VERBOSE: Total number of records for this query: 0

**Reply by emarzanoBNY** [reply]
Popping the last GET into postman returns an empty result.    The question becomes, what is the query needed to find the info via the API? We know it is in there... how do we find it?

**Reply by emarzanoBNY** [reply]
Interesting thing to note... I was feeding the query a child TASK.  When I feed it the parent RITM, the GET query contains a whole bunch of the variables, but the result is still empty.

**Reply by gdbarron** [reply]
When it works properly, the last get above, against sc_item_option_mtom, will return records with the custom vars.  There will then be another call against the source table with sysparm_fields variables.xxxxx.  When giving it the ritm, do you see this?  Can you send over the verbose log for that last part?

**Reply by emarzanoBNY** [reply]
Ah, that was the crux of it!  

I was originally trying this: 

Get-ServiceNowRecord -Table task -IncludeCustomVariable -Filter @('number', '-eq', 'MYTASKNUMBER') 

Updating my command to this: 

Get-ServiceNowRecord -Table sc_req_item -IncludeCustomVariable -Filter @('number', '-eq', 'MYRITMNUMBER') 

Now I've got the variables!

**Reply by gdbarron** [reply]
Awesome!  Glad you got it working.

Do you like the custom vars as is or should we put them under a 'CustomVariable' property?  I'm almost leaning towards another property so they can easily be looped through, but would definitely like your thoughts.

**Reply by emarzanoBNY** [reply]
The custom vars work great as is for me.  I can load the contents of the whole RITM into a single $ticket object.  Once in there, I can grab $ticket.regularVar and $ticket.customVar.  

How would you envision it to work with the CustomVariable property?

**Reply by gdbarron** [reply]
Perhaps $ticket.CustomVariable would be a pscustomobject with properties Name, DisplayName, and Value.  The custom vars would then be separated so they could be easily enumerated and you'd have the benefit of the DisplayName as well.  I'm going to give it some more thought as determining which vars are custom by checking for 'variables.' prepended doesn't feel very user friendly, but perhaps I'm wrong.

**Reply by gdbarron** [reply]
I've updated the branch and created a `CustomVariable` property which looks like
![image](https://user-images.githubusercontent.com/11862024/121735871-bb514b80-cac4-11eb-9f94-b1ab450b1490.png)
If this looks good I'll create a release and get it out there.  Let me know what you think.

**Reply by gdbarron** [reply]
added to version 2.4.1
